### DISCUSSION TITLE
Set up with GPT4Free?

### DISCUSSION THREAD

**Discussion by huachuman** [logancyang/obsidian-copilot#537] [Q&A]
I have been trying to figure out how to set this up with g4f: https://github.com/xtekky/gpt4free

Specifically I'd like to use the Bing provider.

Any help on this would be great. I don't really know where to begin.

**Comment by logancyang** [comment]
@huachuman Looks like you just need to follow their Get Started guide and run it at http://localhost:1337, then in Copilot Settings: OpenAI Proxy Base URL, you put in `http://localhost:1337/v1` then it should work. Let me know if you succeed, if not I can perhaps create a video on this.

**Comment by Harloy** [comment]
Hey there. I'm excited to share how I solved this problem by getting myself a subscription to GPT's, lol.

First, we need a FAKE key. You can find a generator online or use this one: `pysSppnrQyiVhXC5289615F48f0b4f5bAd5aB6C31c442c64`

Then, replace it in the OpenAI Proxy Base URL with `http://localhost:1337/v1`

But that's not all! We need to run the server. Docker is handy, but I'm not great with it. From what I understand, making changes to files through Docker isn't that straightforward. So, let's manually download the g4f repository using git or pip. To use the Copilot plugin, we'll need to change one file.

If you install and run it as-is, the g4f API server will only accept OPTIONS requests instead of POST, and it won't work. To fix this, you need to edit the **__init__.py** file (in my case, it's in user\AppData\Local\Programs\Python\Python310\Lib\site-packages\g4f\api, since I installed it via pip).

Here's what to do:
Add this line at the beginning.
`from fastapi.middleware.cors import CORSMiddleware`
Add this code inside your class:
```
self.app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```
Make sure to keep the indentation correct, it's Python after all. It should look like the screenshot you can find here:
![Code_cNaoXyn58O](https://github.com/logancyang/obsidian-copilot/assets/94390734/dc8c3813-9dad-415b-bd18-df52286b33a1)

Now, you can start the g4f api (by typing those two words in the console).
![image](https://github.com/logancyang/obsidian-copilot/assets/94390734/c83eefcd-d5cc-410d-bc43-5a83aa5018c0)

Voila! Everything should be working smoothly. You're awesome!

By the way, you can also use NeuroGPT. There are no manipulations with the code, but you need to authenticate through Discord, and there are limits on both the free and paid plans (https://neuroapi.gitbook.io/en/)

**Comment by huachuman** [comment]
Is all of that necessary? I think I had gotten it working using @logancyang's post IIRC. But gave up because G4F was not outputting quality responses. Also basically the only working model was Bing and even that wasn't working the same as microsoft's official bing chat.
